File: code_snippets/java/src/main/java/develop/MyKafkaVirtualObject.java
package develop;

import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;

@VirtualObject
public class MyKafkaVirtualObject {

  @Handler
  public void handle(ObjectContext ctx, String req) {
    // <start_headers>
    ctx.request().headers();
    // <end_headers>
  }
}

File: code_snippets/kotlin/src/main/kotlin/develop/MyKafkaVirtualObject.kt
package develop

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.kotlin.ObjectContext

data class MyEvent(val a: Int)

// <start_here>
@VirtualObject
class MyKafkaVirtualObject {

    @Handler
    fun handle(context: ObjectContext, request: MyEvent) {
        // Key
        val kafkaEventKey: String = context.key()

        // ... Do something ...
    }
}
// <end_here>
